lib/AST/TypeUninhabited.cpp

```diff
@@ -0,0 +1,203 @@
+  Inhabitedness checkStruct(CanType type, StructDecl *decl);
+  Inhabitedness checkTuple(TupleType *tuple);
+};
+  
+} // end anonymous namespace
+
+bool TypeBase::isStructurallyUninhabited() {
+  return InhabitantChecker(Type(this)).run() == Uninhabited;
+}
+
+/// The main routine for checking inhabitedness.
+Inhabitedness InhabitantChecker::run() {
+  auto result = checkOrPush(originalType);
+  if (result != Unknown) {
+    return result;
+  }
+  
+  while (!stack.empty()) {
+    auto type = stack.back();
+    auto result = checkType(type);
+    if (result != Unknown) {
+      stack.pop_back();
+      isInhabited.insert({type, result == Inhabited});
+    }
+  }
+  
+  auto it = isInhabited.find(originalType->getCanonicalType());
+  assert(it != isInhabited.end());
+  return it->second ? Inhabited : Uninhabited;
+}
+
+Inhabitedness InhabitantChecker::checkOrPush(Type type) {
+  auto canonical = type->getCanonicalType();
+  
+  if (canonical->isUninhabited())
+    return Uninhabited;
+  
+  if (auto tuple = dyn_cast<TupleType>(canonical))
+    return checkTuple(tuple);
+  
+  auto nominal = canonical.getAnyNominal();
+  if (!nominal)
+    return Inhabited;
+  
+  auto params = nominal->getGenericParams();
+  // Only bother with types that become inhabited because of their generic parameters.
+  // If the type is declared to be inhabited (but isn't an enum--that's detected above)
+  // then it'll generate a warning anyway.
+  if (!params)
+    return Inhabited;
+  
+  // If the type has a cycle, bail and call in inhabited.
+  if (std::find(stack.begin(), stack.end(), canonical) != stack.end())
+    return Inhabited;
+  
+  auto it = isInhabited.find(canonical);
+  if (it != isInhabited.end())
+    return it->second ? Inhabited : Uninhabited;
+
+  stack.push_back(canonical);
+  return Unknown;
+}
+
+Inhabitedness InhabitantChecker::checkType(CanType type) {
+  if (auto decl = type.getAnyNominal()) {
+    if (auto c = dyn_cast<ClassDecl>(decl)) {
+      return checkClass(type, c);
+    } else if (auto e = dyn_cast<EnumDecl>(decl)) {
+      return checkEnum(type, e);
+    } else if (auto s = dyn_cast<StructDecl>(decl)) {
+      return checkStruct(type, s);
+    } else {
+      return Inhabited;
+    }
+  } else if (auto tuple = dyn_cast<TupleType>(type)) {
+    return checkTuple(tuple);
+  } else {
+    return Inhabited;
+  }
+}
+
+Inhabitedness InhabitantChecker::checkEnum(CanType type, EnumDecl *decl) {
+  for (auto elt : decl->getAllElements()) {
+    switch (checkEnumCase(elt)) {
+      case Unknown:
+        return Unknown;
+      case Inhabited:
+        return Inhabited;
+      case Uninhabited:
+        break;
+    }
+  }
+  return Uninhabited;
+}
+
+Inhabitedness InhabitantChecker::checkEnumCase(EnumElementDecl *decl) {
+  auto params = decl->getParameterList();
+  for (auto i = params->begin(); i != params->end(); ++i) {
+    auto *param = *i;
+    auto paramType = param->getType();
```

Comment:
Consider these types:

```swift
enum T4<A> {
	case x(A)
}
enum T5<X> {
	case x(T4<X>)
}
```

In the case where the enum is specialized (`T5<Never>`), then its type is a bound generic that's specialized to `Never`. (Is _specialized_ the right word there?)

But when I get the parameters, they're not specialized. So this is `T4<X>`, not `T4<Never>`. Is there a way to get the latter?